sigstore-java/src/test/java/dev/sigstore/tuf/TufClientTest.java

```diff
@@ -146,6 +148,97 @@ public void testRootUpdate_metaFileTooBig()
     }
   }
 
+  @Test
+  public void testTimestampUpdate_throwMetaNotFoundException() throws IOException {
+    setupMirror("remote-timestamp-not-present");
+    var client = createTimeStaticTufClient(localStore);
+    assertThrows(
+        MetaNotFoundException.class,
+        () -> {
+          client.updateTimestamp();
+        });
+  }
+
+  @Test
+  public void testTimestampUpdate_throwsSignatureVerificationException()
+      throws IOException, NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException {
+    setupMirror("remote-timestamp-unsigned", "2.root.json", "3.root.json", "timestamp.json");
+    var client = createTimeStaticTufClient(localStore);
+    try {
+      client.updateRoot();
+      client.updateTimestamp();
+      fail();
```

Comment:
Frankly speaking, it would be nice to have `assertThrows` in such cases.
The bad thing with standalone `fail()` is that the test failure would be extremely obscure to a new developer. They would have to reverse-engineer the reasoning behind the test code.

If you add a relevant message, then the test failure would look like a decent bug report:

```java
assertThrows(
    SignatureVerificationException.class,
    () -> { test code },
    () -> "root signature is missing in ..., so updateRoot should fail"
);
```

The same goes for `assertEquals(0, e.getVerifiedSignatures());`
It produces unclear error messages.

Reply:
Hey Vladimir,

I agree with you there's room for improvement here. I forgot to add error messages in these cases so I've gone and added them. I can't say I'm a fan of the assertThrows approach for cases where I want to do extra verification of the exception contents. It just seems to make the code more deeply nested and harder to read.  I hope the messages I've included improve things somewhat.

Reply:
> It just seems to make the code more deeply nested and harder to read

Well, if you supply the expected message via `assertThrows`, then it would be printed even in the case the actual exception has a different class.

> It just seems to make the code more deeply nested and harder to read

Frankly speaking,

```java
    var exception = assertThrows(
      SignatureVerificationException.class,
      () -> {
        client.updateRoot();
        client.updateTimestamp();
      },
      "The timestamp was not signed so should have thown a SignatureVerificationException"
    );
    assertAll(
      () -> assertEquals(0, exception.getVerifiedSignatures()),
      () -> assertEquals(1, exception.getRequiredSignatures())
    );
```

looks pretty much the same to me as

```java
    try {
      client.updateRoot();
      client.updateTimestamp();
      fail("The timestamp was not signed so should have thown a SignatureVerificationException.");
    } catch (SignatureVerificationException e) {
      assertEquals(0, e.getVerifiedSignatures());
      assertEquals(1, e.getRequiredSignatures());
    }
```

Even though `try-catch` variation might look "easier to read", I believe, people rarely read tests. The test either just lives on its own, or you read **test failure**. In any case, that is, well, Java :)

Yet another possibility is to prefer `AssertJ` which provides better API and better error messages: https://www.baeldung.com/assertj-exception-assertion#1-using-assertthatthrownby

Truth yields better messages, yet it is not very compatible with JUnit 5.

---

At the same time, I don't understand  TUF tests and code, so if think `assertThrows` does not suit here, ignore it. However, I'm inclined  `assertThrows` is not that bad. However, I'm not going to veto `try-case`-based tests.

PS. The same in Kotlin would be:

```kotlin
import org.junit.jupiter.api.assertThrows // <-- out of the box JUnit

        val exception = assertThrows<SignatureVerificationException>(
            "The timestamp was not signed so should have thown a SignatureVerificationException"
        ) {
            client.updateRoot()
            client.updateTimestamp()
        }
        assertAll(
            { assertEquals(0, exception.verifiedSignatures) },
            { assertEquals(1, exception.requiredSignatures) },
        )
```